vala: Building with flat layout does not work

It does not seem to resolve the generated header path properly:

```
[5/7] Compiling c object 'meson-out/valaprog@exe/meson-out_valaprog@exe_prog.c.o'
FAILED: meson-out/valaprog@exe/meson-out_valaprog@exe_prog.c.o 
cc  '-Imeson-out/valaprog@exe' '-fdiagnostics-color=always' '-I../prog' '-Iprog' '-Ilib' '-I../lib' '-pipe' '-w' '-O0' '-g' '-I/usr/include/glib-2.0' '-I/usr/lib64/glib-2.0/include' '-MMD' '-MQ' 'meson-out/valaprog@exe/meson-out_valaprog@exe_prog.c.o' '-MF' 'meson-out/valaprog@exe/meson-out_valaprog@exe_prog.c.o.d' -o 'meson-out/valaprog@exe/meson-out_valaprog@exe_prog.c.o' -c 'meson-out/valaprog@exe/prog.c'
meson-out/valaprog@exe/prog.c:9:21: erreur fatale : valalib.h : No such file or directory
 #include "valalib.h"
                     ^
compilation terminée.
[6/7] Generating symbol file 'meson-out/valalib@sha/meson-out/libvalalib.so.symbols'.
ninja: build stopped: subcommand failed.
```

Also, there should be some CI support to run all the test cases using the flat layout.

I'll probably look into this.

--- a/mesonbuild/msetup.py
+++ b/mesonbuild/msetup.py
@@ -177,8 +177,6 @@ class MesonApp:
         mlog.initialize(env.get_log_dir(), self.options.fatal_warnings)
         if self.options.profile:
             mlog.set_timestamp_start(time.monotonic())
-        if env.coredata.options[mesonlib.OptionKey('backend')].value == 'xcode':
-            mlog.warning('xcode backend is currently unmaintained, patches welcome')
         with mesonlib.BuildDirLock(self.build_dir):
             self._generate(env)
 
@@ -264,6 +262,11 @@ class MesonApp:
 
             # Post-conf scripts must be run after writing coredata or else introspection fails.
             intr.backend.run_postconf_scripts()
+
+            # collect warnings about unsupported build configurations; must be done after full arg processing
+            # by Interpreter() init, but this is most visible at the end
+            if env.coredata.options[mesonlib.OptionKey('backend')].value == 'xcode':
+                mlog.warning('xcode backend is currently unmaintained, patches welcome')
         except Exception as e:
             mintro.write_meson_info_file(b, [e])
             if 'cdf' in locals():

--- a/mesonbuild/msetup.py
+++ b/mesonbuild/msetup.py
@@ -267,6 +267,13 @@ class MesonApp:
             # by Interpreter() init, but this is most visible at the end
             if env.coredata.options[mesonlib.OptionKey('backend')].value == 'xcode':
                 mlog.warning('xcode backend is currently unmaintained, patches welcome')
+            if env.coredata.options[mesonlib.OptionKey('layout')].value == 'flat':
+                mlog.warning('-Dlayout=flat is unsupported and probably broken. It was a failed experiment at '
+                             'making Windows build artifacts runnable while uninstalled, due to PATH considerations, '
+                             'but was untested by CI and anyways breaks reasonable use of conflicting targets in different subdirs. '
+                             'Please consider using `meson devenv` instead. See https://github.com/mesonbuild/meson/pull/9243 '
+                             'for details.')
+
         except Exception as e:
             mintro.write_meson_info_file(b, [e])
             if 'cdf' in locals():

